### PR TITLE
added: better fuzzy matching

### LINKED ISSUE #2095: conflict between icon for i3wm workspace label "1" and "10" when display by polybar
## Describe the issue

**Expected behavior:**
I've ten workspace configure in i3 config labeled from 1 to 10.
In polybar config, for i3 module I use `ws-icon-[0-9]+ = <label>;<icon>` to have different icon for most of my workspace.
Concerning the bug, I should have a different icon for each 1st and 10th workspace.

**Actual behavior:**
The 10th workspace takes the 1st workspace icons instead of the of the one configure for it.

**Was it working before?**
* Did you also experience this bug in an earlier version of polybar? **No.**
* If no, what was the last version where this worked correctly? **First time I use Polybar. I'm finishing setting it up**

**How I configure workspaces in i3wm:**
I use variables to define them :
```
set $ws1 "1"
set $ws2 "2"
...
set $ws10 "10"
```
then I affect each one to different keybinding, for example with the first :
```
bindsym $mod+ampersand workspace number $ws1r
```

**Then in polybar module config:**
```
ws-icon-0 = 1;  #This is Font Awesome icon
ws-icon-1 = 2;
ws-icon-2 = 3;
ws-icon-3 = 4;
ws-icon-4 = 5;
ws-icon-5 = 6;
ws-icon-6 = 7;
ws-icon-7 = 8;
ws-icon-8 = 9;
ws-icon-9 = 10;
```

**I've tried two things to see if there was a problem with the icon or if I did something wrong:**

- try `ws-icon-9 = "10";` instead of `ws-icon-9 = 10;` -> no change

- try to change the label in both i3 config and polybar config :
I use `set $ws10 "ten"` and `ws-icon-9 = ten;`. I could have the right logo (opera logo) but with a -1 before (I suppose because the label is normally a number)
see the screenshots

## Screenshots
Capture d’écran_2020-05-10_23-52-34
Capture d’écran_2020-05-10_23-54-02

(It's a two screen setup)

**With the i3 module biger:**
Capture d’écran_2020-05-10_23-57-01     Capture d’écran_2020-05-10_23-58-18

**With the good icon but the wrong number:**
Capture d’écran_2020-05-11_00-07-32

**With what I see, I think it could be something I'm doing wrong or a bug in polybar (?)**
What I don't understand is why with "ten" it display the right icon but it doesn't with "10"
If you need something else, ask me.
I don't know if you need logs of polybar since it don't seems to be an error.. I'll look logs tomorrow to see if there is any Warnings just in case (It's getting late in France ;) )

## Environment:
* WM:i3 version 4.18.1 (2020-04-23) © 2009 Michael Stapelberg and contributors
* Distro:

DISTRIB_ID=ManjaroLinux
DISTRIB_RELEASE=20.0
DISTRIB_CODENAME=Lysia
DISTRIB_DESCRIPTION="Manjaro Linux

* Output of `polybar -vvv`:

polybar 3.4.2

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) +composite +xkb +xrm +xcursor

Build type: Release
Compiler: /usr/bin/c++
Compiler flags: -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wall -Wextra -Wpedantic -O3 -DNDEBUG
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now

### LINKED ISSUE #2829: [Feature]: Better fuzzy matching for icon sets (i3, bspwm)
### Checklist

- [X] I have read the appropriate section in the [contributing guidelines](https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md)
- [X] I believe this issue is a problem with polybar itself and not a misconfiguration on my part
- [X] I have searched for other open and closed [issues](https://github.com/polybar/polybar/issues?q=is%3Aissue) that may have already reported this problem
- [X] I have checked the [known issues](https://github.com/polybar/polybar/wiki/Known-Issues) page for this problem.
- [X] I have followed the [debugging guide](https://github.com/polybar/polybar/wiki/Debugging-your-Config) to narrow down the problem to a minimal config.

### Steps to reproduce

1. run a polybar config with the i3 module
2. include ws-icon layout from the minimal config
3. conflict

### Minimal config

```dosini
[module/i3]
type = internal/i3

pin-workspaces = true
strip-wsnumbers = true
index-sort = true
enable-click = true
enable-scroll = false
wrapping-scroll = false
reverse-scroll = false

fuzzy-match = true

ws-icon-0 = "1;"
ws-icon-1 = "2;"
ws-icon-2 = "3;"
ws-icon-3 = "4;"
ws-icon-4 = "5;"
ws-icon-5 = "6;6"
ws-icon-6 = "7;"
ws-icon-7 = "8;"
ws-icon-8 = "9;"
ws-icon-9 = "10;💬"
ws-icon-default = 💬

format = <label-state> <label-mode>
```

### Polybar log

```text
no warnings or errors
```

### Expected behavior

the icon for workspace 10 should be 💬.

### Actual behavior

The icon for workspace 10 is the same as the first one, which is ws-icon-0

### Window Manager and Version

i3-gaps

### Linux Distribution

EndeavourOS

### Polybar version

```text
polybar 3.6.3

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) +composite +xkb +xrm +xcursor

Build type: Release
Compiler: /usr/bin/c++
Compiler flags: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -g -ffile-prefix-map=/build/polybar/src=/usr/src/debug -flto=auto -O3 -DNDEBUG -Wall -Wextra -Wpedantic -Wsuggest-override
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -flto=auto -Wall -Wextra -Wpedantic -Wsuggest-override  -Wall -Wextra -Wpedantic -Wsuggest-override
```

### Additional Context / Screenshots

image

This is the same bug as https://github.com/polybar/polybar/issues/2095, which appears to have already been resolved, but, as mentioned by others is still sometimes occurring. 

Thank you!

### PR DESCRIPTION
**Pull request by ron0studios**
Improved fuzzy matching so that it finds the longest matching icon instead of the first possible match

## What type of PR is this? (check all applicable)

* [ ] Refactor
* [x] Feature
* [x] Bug Fix
* [ ] Optimization
* [ ] Documentation Update
* [ ] Other: *Replace this with a description of the type of this PR*

## Description

This improves the `fuzzy-match` option in a few modules, such as the i3 module. Instead of matching the first applicable substring, the option will now search through and find the longest matching substring.

## Related Issues & Documents

Closes #2829
Fixes #2095 (similar issue)

## Documentation (check all applicable)

* [x] This PR requires changes to the Wiki documentation (describe the changes)
* [ ] This PR requires changes to the documentation inside the git repo (please add them to the PR).
* [ ] Does not require documentation changes

### CODE REVIEW

**Comment by ron0studios** [comment]
please let me know if I've forgotten anything, this is my first PR 😄

**Comment by ron0studios** [comment]
just removed the debug output, should fix the workflow check fails

File: include/drawtypes/iconset.hpp [hunk 1]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <map>
+#include <algorithm>
```
Comment (patrick96) [inline include/drawtypes/iconset.hpp:4]: Can (and should) be moved to the cpp file

File: src/drawtypes/iconset.cpp [hunk 2]
```diff
@@ -20,9 +21,11 @@ namespace drawtypes {
 
     // If fuzzy matching is turned on, try that first before returning the fallback.
     if (fuzzy_match) {
+      // works by finding the *longest* matching icon to the given workspace id
+      int max_size = -1;
       for (auto const& icon : m_icons) {
         if (id.find(icon.first) != std::string::npos) {
-          return icon.second;
+          max_size = std::max(max_size, (int)icon.first.length());
```
Comment (patrick96) [inline src/drawtypes/iconset.cpp:28]: You're finding the length of the longest match, but you're not actually doing anything with it.

Comment (patrick96) [reply src/drawtypes/iconset.cpp:28]: You can also write some tests in `tests/unit_tests/drawtypes/iconset.cpp` to check that the function actually does what you want it to do ([testing documentation](https://polybar.readthedocs.io/en/latest/dev/testing.html))

File: include/drawtypes/iconset.hpp (hunk 1 above)
Comment (ron0studios) [reply include/drawtypes/iconset.hpp:4]: moved to iconset.cpp

**Review by ron0studios — COMMENTED** [review]
just made some more commits with the given changes

**Review by ron0studios — COMMENTED** [review]
fully finished implementing all changes in the recent commit

**Review by ron0studios — COMMENTED** [review]
resolved

File: src/drawtypes/iconset.cpp [hunk 3]
```diff
@@ -20,11 +21,21 @@ namespace drawtypes {
 
     // If fuzzy matching is turned on, try that first before returning the fallback.
     if (fuzzy_match) {
+      // works by finding the *longest* matching icon to the given workspace id
+      int max_size = -1;
```
Comment (patrick96) [inline src/drawtypes/iconset.cpp:25]: If you make this have type `size_t`, you don't have to cast to `int` further down.

You will also have to change the condition under which you update `max_size` and `max_label` by adding `|| !max_label`. This way, the first match will update `max_size`.

**Review by patrick96 — CHANGES_REQUESTED** [review]
Only some minor style changes. Please also run `clang-format` on the files you changed. https://polybar.readthedocs.io/en/latest/dev/style-guide.html

File: include/drawtypes/iconset.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ron0studios) [reply include/drawtypes/iconset.hpp:4]: woops forgot to save my hpp file sorry!

File: src/drawtypes/iconset.cpp [hunk 4]
```diff
@@ -20,11 +21,21 @@ namespace drawtypes {
...
+      label_t max_label = NULL;
```
[the remark this answers is not in the document]
Comment (ron0studios) [reply src/drawtypes/iconset.cpp]: made the suggested change

File: src/drawtypes/iconset.cpp [hunk 5]
```diff
@@ -20,11 +21,21 @@ namespace drawtypes {
...
-          return icon.second;
+
+          if((int)icon.first.length() > max_size) {
+            max_size = icon.first.length();
+            max_label = icon.second;
+          }
         }
       }
+
+      if(max_size != -1) return max_label;
```
[the remark this answers is not in the document]
Comment (ron0studios) [reply src/drawtypes/iconset.cpp]: apologies

File: src/drawtypes/iconset.cpp (hunk 3 above)
Comment (ron0studios) [reply src/drawtypes/iconset.cpp]: just made the necessary changes for this, thank you

**Comment by ron0studios** [comment]
there is still a pending requested change, but don't worry as I've just implemented those changes myself and the conversation was made outdated. All the relevant changes have been applied. Apologies for all the mistakes again, thank you!

**Comment by ron0studios** [comment]
oh and I forgot to mention that I am making this PR as part of hacktoberfest

File: src/drawtypes/iconset.cpp [hunk 6]
```diff
@@ -20,11 +22,22 @@ namespace drawtypes {
...
         if (id.find(icon.first) != std::string::npos) {
-          return icon.second;
+          if (icon.first.length() > max_size || !max_label) {
+            max_size = icon.first.length();
+            max_label = icon.second;
+          }
         }
       }
+
+      if (max_size != -1) {
```
[the remark this answers is not in the document]
Comment (ron0studios) [reply src/drawtypes/iconset.cpp:38]: Ill change that to `if(max_label)` instead

File: .vscode/tasks.json [hunk 7]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (patrick96) [inline .vscode/tasks.json:1]: That file is not needed ;)

Comment (ron0studios) [reply .vscode/tasks.json:1]: 🤦🏽‍♂️

**Review by patrick96 — APPROVED** [review]

**Comment by patrick96** [comment]
Congrats on your first PR in polybar :rocket:

**Pull request merged**